### PR TITLE
add Reqd.schedule_trailers (fixes #97)

### LINKED ISSUE #97: Trailing headers
It seems impossible to use this library to sent

HEADERS frame
DATA frame
HEADERS frame

This is something that would be needed to implement GRPC for ocaml (https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md)

### PR DESCRIPTION
**Pull request by blandinw**
This PR makes sending trailing headers (aka. trailers) possible, and thus allows H2 to be used to implement gRPC.
I've been going back and forth on the best patch to write, and I believe this strikes a good balance, but I'm curious to hear your thoughts.
I had a look at the changes proposed in #72 and I believe it suffers from the sequencing issue described below (commit #1).
However, I did steal the test case from there.

Commit 1 (move Streaming flush logic out of Body):
- Body.close_writer caused stream to close before we had the chance to send headers
- Body.close_writer caused next flush to send a zero length DATA frame, which leads to a sequence like HEADERS, DATA, TRAILERS, DATA
- imo, Body should not be concerned with HTTP details (knowing about RFC7540 and sending DATA frames)
- moving the corresponding code out of Body and into Reqd feels like the right thing to do since it allows for better sequencing + simpler code (`has_pending_output` forwards to Faraday, no more `t.write_final_data_frame` and no more `response_body_requires_output`)
I also noticed a potential bug where I believe the zero length DATA frame is never sent if we're flow-controlled (maxxed out sent bytes) in `Scheduler.write`

Commit 2 (allow trailing headers, fixes #97)
- straight-forward thanks to the sequencing control allowed by commit 1
- we choose not to let user send trailers directly, but instead register them and later send them on Body.close to avoid user mistakes (sending trailers too early or too late)
- this also allows to avoid sending a useless zero length DATA frame before the trailers

Quick note: I was confused at the beginning as to how the state machine worked, until I realized it relied on Gluten. It's not obvious at first if you're grepping around (could not find `next_write_operation` call sites)
You may want to have a quick blurb or schema showing the `Gluten -> Server_connection -> Scheduler -> Reqd -> Writer -> Faraday` chain.

### CODE REVIEW

**Comment by anmonteiro** [comment]
Thanks for sending this. I'll try to get to it this weekend.

Before I start reviewing, a potentially related question: would gRPC also need to read trailer headers, in addition to sending them (what this PR does)? There's some machinery in place for that but it's not fully wired up.

**Comment by blandinw** [comment]
Hi Antonio, thanks!

Clients need the ability to read the response trailers since they contain the actual status code (HTTP code should always be 200, regardless of errors), and an optional message. Clients do not need the ability to send request trailers as far as I know.

**Comment by anmonteiro** [comment]
> I also noticed a potential bug where I believe the zero length DATA frame is never sent if we're flow-controlled (maxxed out sent bytes) in Scheduler.write

@blandinw I think you're right about this. If you don't mind, I'm going to push a test + a fix for that before discussing this PR further.

File: lib/h2.mli
Comment (anmonteiro) [inline lib/h2.mli:505]: nit: I'd prefer if this was rather called `schedule_trailers` or `schedule_trailer_headers`

Comment (anmonteiro) [reply lib/h2.mli:505]: Can you also add a docstring, please?

File: lib/reqd.ml
Comment (anmonteiro) [inline lib/reqd.ml:54]: I think a triple is probably the cutoff in my mind for preferring a record.

I'd rather see a record (can probably be inline) here, and I think we don't necessarily need the type of trailers to be optional. Initializing it with `Headers.empty` and comparing to that will probably save allocations.

File: lib/reqd.ml
Comment (anmonteiro) [inline lib/reqd.ml:198]: thoughts on failing if you try to schedule multiple times? I think this behavior should be somewhat specified, and additionally documented in the `.mli`

File: lib/reqd.ml
Comment (anmonteiro) [inline lib/reqd.ml:207]: I think it's possible, in certain cases, to retain a reference to the request descriptor when the stream has been closed (e.g. because of an error). The other cases seem fine to me, but the `Closed` case should probably be handled separately.

File: lib/reqd.ml
Comment (anmonteiro) [inline lib/reqd.ml:216]: I think this doesn't apply to trailer headers.

File: lib/reqd.ml
Comment (anmonteiro) [inline lib/reqd.ml:501]: Why does this need to be a function? it looks like you're always setting the `END_STREAM` flag below, anyway?

File: lib/reqd.ml
Comment (anmonteiro) [inline lib/reqd.ml:504]: Similary, this is only used once below. I'd rather see it inlined

File: lib/serialize.ml
Comment (anmonteiro) [inline lib/serialize.ml:547]: Note to self: we could probably unify this code with the request / response header serializers above.

File: lib_test/test_h2_server.ml
Comment (anmonteiro) [inline lib_test/test_h2_server.ml:1256]: remove `TODO` please :)

File: lib_test/test_h2_server.ml
Comment (anmonteiro) [inline lib_test/test_h2_server.ml:1249]: nit: there seems to be a remainder of an OCaml comment in this string?

File: lib_test/test_h2_server.ml
Comment (anmonteiro) [inline lib_test/test_h2_server.ml:1219]: it would be cool to verify that this `DATA` frame doesn't have the `END_STREAM` flag set.

File: lib_test/test_h2_server.ml
Comment (anmonteiro) [inline lib_test/test_h2_server.ml:1233]: Similarly, it'd be cool to verify that this `HEADERS` frame _does_ have the `END_STREAM` flag set.

**Review by anmonteiro — COMMENTED** [review]
@blandinw I like the direction of this PR and the approach you took. Thanks again for sending it over!

I only left mostly minor stylistic / documentation comments. Additionally, would you mind adding an entry to the `CHANGES.md` file?

Note: I force-pushed to this branch after merging #142, so please make sure to reset locally.

File: lib/reqd.ml
Comment (blandinw) [reply lib/reqd.ml:198]: I can't think of a use case where we'd need to change the trailers after setting them, and we should notify the user somehow that something is probably wrong.

I was hoping we could somehow avoid runtime errors, but I'm not sure how to do it with the current API. Maybe something to think about if we decide to change the API one day (we could then probably enforce it at the type level).

I'll error out for now.

**Review by blandinw — COMMENTED** [review]
Addressed your comments in a new commit for easier review, please squash before merging

File: lib/reqd.ml
Comment (anmonteiro) [inline lib/reqd.ml:513]: the previous behavior flushed the writer before marking the stream as closed. Was this change intentional?

**Review by anmonteiro — APPROVED** [review]
@blandinw thank you for addressing my comments. this looks almost ready now! I just left one additional question.

File: lib/reqd.ml
Comment (blandinw) [reply lib/reqd.ml:513]: Here's the recent history for this line of code
- I added the `flush` call with a `Stream.finish_stream` in https://github.com/anmonteiro/ocaml-h2/commit/d7e57959930e669bd400a80fec36ba7c72413b76#diff-71bafa60481b1abadc94a8c1599b45c6R492 (tests would fail otherwise)
- you made the `Stream.finish_stream` a `close_stream` call in https://github.com/anmonteiro/ocaml-h2/commit/d7e57959930e669bd400a80fec36ba7c72413b76#diff-71bafa60481b1abadc94a8c1599b45c6R492
- since `close_stream` already flushes (either directly, or via `reset_stream`, I removed the explicit flush call

`dune runtest` and my toy gRPC project are running fine, and a quick look at the code makes me think it should work, but I'll defer to you since you're more experienced with the codebase.

Comment (anmonteiro) [reply lib/reqd.ml:513]: Good point -- I forgot `close_stream` already flushed.

**Comment by anmonteiro** [comment]
Thanks again!

**Comment by blandinw** [comment]
Happy to contribute, it was fun 😄

**Pull request merged**
